src/lib/mdns/minimal/responders/QueryResponder.h

```diff
@@ -0,0 +1,225 @@
+    {
+        if (mRemaining != 0)
+        {
+            mCurrent++;
+            mRemaining--;
+        }
+        SkipInvalid();
+        return *this;
+    }
+
+    QueryResponderIterator operator++(int)
+    {
+        QueryResponderIterator tmp(*this);
+        operator++();
+        return tmp;
+    }
+
+    bool operator==(const QueryResponderIterator & rhs) const { return mCurrent == rhs.mCurrent; }
+    bool operator!=(const QueryResponderIterator & rhs) const { return mCurrent != rhs.mCurrent; }
+
+    QueryResponderRecord & operator*() { return *mCurrent; }
+    QueryResponderRecord * operator->() { return mCurrent; }
+
+    Internal::QueryResponderInfo * GetInternal() { return mCurrent; }
+    const Internal::QueryResponderInfo * GetInternal() const { return mCurrent; }
+
+private:
+    bool isCurrentValid() const
+    {
+        if (mCurrent->responder == nullptr)
+        {
+            return false;
+        }
+
+        if (mAdditionalOnly && !mCurrent->reportNowAsAdditional)
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    /// Skips invalid/not useful values.
+    /// ensures that if mRemaining is 0, mCurrent is nullptr;
+    void SkipInvalid()
+    {
+        while ((mRemaining > 0) && !isCurrentValid())
+        {
+            mRemaining--;
+            mCurrent++;
+        }
+        if (mRemaining == 0)
+        {
+            mCurrent = nullptr;
+        }
+    }
+
+    bool mAdditionalOnly;
+    Internal::QueryResponderInfo * mCurrent;
+    size_t mRemaining;
+};
+
+/// Responds to mDNS queries.
+///
+/// In particular:
+///   - replies data as provided by the underlying responders
+///   - replies to "_services._dns-sd._udp.local."
+///
+/// Maintains a stateful list of 'additional replies' that can be marked/unmarked
+/// for query processing
+class QueryResponderBase : public Responder // "_services._dns-sd._udp.local"
+{
+public:
+    /// Builds a new responder with the given storage for the response infos
+    QueryResponderBase(Internal::QueryResponderInfo * infos, size_t infoSizes);
+    virtual ~QueryResponderBase() {}
+
+    // Setup initial settings (clears all infos and sets up dns-sd query replies)
+    void Init();
+
+    /// Add a new responder to be processed
+    ///
+    /// Return true if add was succesful.
+    QueryResponderSettings AddResponder(Responder * responder);
+
+    /// Implementation ofthe responder delegate
+    ///
+    /// adds responses for all known _dns-sd services
+    void AddAllResponses(const chip::Inet::IPPacketInfo * source, ResponderDelegate * delegate) override;
+
+    QueryResponderIterator begin() { return QueryResponderIterator(false, mResponderInfos, mResponderInfoSize); }
+    QueryResponderIterator end() { return QueryResponderIterator(); }
+
+    // Keeping track of 'additional data to report'
+
+    void ResetAdditionals();
+
+    /// Marks queries matching this qname as 'to be additionally reported'
```

Comment:
Overall, I'm having trouble understanding the contract of all API where there is "to be additionally reported". Is that to fulfill Section 12 of RFC6763 (https://tools.ietf.org/html/rfc6763#section-12) i.e. PTR/SRV records should have A/AAAA ?

Reply:
Let me know how to rephrase this, but generally I have replies containing 'answers', 'authoritative' and 'additional'.

I add any direct answers to the query in 'answers', don't use authoritative and add in 'additional' anything 'linked.

So if I add a PTR record towards foo, I will mark 'foo' as 'report additional'
Then if foo is a SRV record pointing towards bar, then I mark this 'bar' as additional for SRV
Registation looks like this logically:

```
xyz.local:   PTR foo.local  // mark foo.local as additional
foo.local:   SRV 1234 0 0 bar.local  // mark bar.local as additional
bar.local:  A  1.2.3.4
bar.local: AAAA <someip>
```

then when a query of xyz.local/ANY comes in, the minMdns code will say:

Answer:  xyz.loca/PTR/foo.local   // because this answers the query
Additional:  foo.local/SRV/1234/0/0/bar.local  // because I am asked to return foo.local when I answer the PTR
Additional: bar.local/A/1.2.34   // because I am asked to return bar.local when I answer the SRV
Additional: bar.local/AAAAA/<someip>  // because I am asked to return bar.local when I answer the SRV

This is how I saw other libs behave (chromecast, dns-sd, the esp32 sample when it worked).

I am not sure if the logic is that sane regarding all cases, it just happens to do the right thing for the cases I cared about and it made sense to me (when a record references something else, report the 'something else' as well)